### ISSUE TITLE
Support "Sign In with Apple" auth [iOS]

### ISSUE THREAD

**Issue by gnprice** [zulip/zulip-mobile#3964]
This is for the client side, on iOS, of zulip/zulip#14168.

In addition to the links there, here are Apple docs specific to iOS:
* https://developer.apple.com/documentation/authenticationservices
* sample code: https://developer.apple.com/documentation/authenticationservices/implementing_user_authentication_with_sign_in_with_apple

Here's an implementation for RN which we may be able to use:
* https://github.com/expo/expo/tree/master/packages/expo-apple-authentication

Note that although it's for React Native, it only works on iOS. It's a thin wrapper around the system-provided components.

We'll also want to support this on Android, because people may sign up using it on iOS and then also want to use Zulip on an Android device. That may have a separate implementation, so I'll file a separate issue for it.

**Comment by mateuszmandera** [comment]
I think this one should be possible to work on without having anything done on the server yet? The first part of the native flow should be happening only between the app and Apple server - at the end of that the app should have a JWT token. All that's left at that point is to send the token with `mobile_flow_otp` to a server endpoint (probably `/complete/apple-id`) and the server will give the standard `zulip://` redirect response - testing this part will require an implementation on the server.

**Comment by gnprice** [comment]
Hmm -- yeah, after looking into this side of things some more, I agree. There's enough complexity involved before the point where we talk to the server that we can productively start working through that even while the server implementation is still being prepared.

OTOH, there is a different prerequisite this is blocked on! That handy existing implementation for RN requires something called [react-native-unimodules](https://github.com/unimodules/react-native-unimodules), which in turn requires us to set up CocoaPods for our build. That's something we've been meaning to do, but will take some work to accomplish. I've just filed #3983 to give it an issue of its own; now that it's blocking this work, we should get on that soon.

So, thanks @mateuszmandera for poking us on this!

**Comment by chrisbobbe** [comment]
@gnprice, when you get a chance, there are some necessary Apple Developer Console actions that you're authorized to take. From Expo's AppleAuthentication doc ([link to "Configuration" section](https://docs.expo.io/versions/latest/sdk/apple-authentication/#configuration)):

> 2. Log into the Apple Developer Console, go to "Certificates, Identifiers, & Profiles" and then "Identifiers".
> [...]
> 4. In the list of identifiers, click on the one corresponding to your primary app. Enable the "Sign In with Apple" capability, click "Edit", and choose the "Enable as a primary App ID" option. Save the new configuration.
> [...]
> 6. Next, go to the "Keys" page and register a new key. Add the "Sign In with Apple" capability, and make sure to choose the correct primary app on the configuration screen.

**Comment by gnprice** [comment]
> @gnprice, when you get a chance, there are some necessary Apple Developer Console actions that you're authorized to take.

Thanks -- done!

Let me know if you need the Key ID or other details and can't see them in the console yourself.

The key itself, I imagine will end up on our production servers to be actually used, as part of deploying zulip/zulip#14168. I've saved it locally, and also sent a copy to Tim (end-to-end encrypted) so there's a backup.

**Comment by chrisbobbe** [comment]
OK, sounds good!

We do have some control over the appearance of the button, but not enough to make it look just like the other buttons. ([Apple doc](https://developer.apple.com/documentation/authenticationservices/asauthorizationappleidbutton), [Expo doc](https://docs.expo.io/versions/latest/sdk/apple-authentication/#appleauthenticationappleauthenticationbutton))

It looks like we have full control over the corner radius, as long as we use a special prop, which is fine. And we can set the height and width to match the other buttons.

At that point, we can set the following:

For `buttonType`:
- `SIGN_IN`: The button's text (in English, anyway) is "Sign in with Apple".
- `CONTINUE`: The button's text (in English, anyway) is "Continue with Apple". Here's one screenshot of `CONTINUE`:

![image](https://user-images.githubusercontent.com/22248748/78415609-9d8e7980-75d7-11ea-92ac-f984ea8fb0dc.png)

For `buttonStyle`:
- `WHITE`,
- `WHITE_OUTLINE`
- `BLACK`.

![apple sign in button](https://user-images.githubusercontent.com/22248748/78415615-aa12d200-75d7-11ea-8c1b-240b4780776a.png)

**Comment by gnprice** [comment]
Hmmm, interesting.

Those Expo docs say some things that aren't true, like:
> The App Store Guidelines require you to use this component to start the authentication process instead of a custom button.

when here's [Apple's guidelines](https://developer.apple.com/design/human-interface-guidelines/sign-in-with-apple/overview/buttons/) on using a custom button:
> If your layout requires it, you can create a custom Sign in with Apple button for iOS, macOS, or the web. For example, if you support multiple sign-in methods, you may want to display sign-in buttons that use left-aligned logos, or that display a logo only. [...]

That decreases my confidence somewhat in the accuracy of this Expo library's implementation. :-/

Still, surely going to be easiest at least for now to go along with that library's limitations, even if it means things look a little janky.

* For where we say "Log in", the right text is SIGN_IN.
* I think I prefer the button styles that look most like the other buttons.
  * A possible followup is to take this as a prompt to revisit the other buttons' designs too -- even apart from consistency with the Apple buttons, they can probably be improved.

**Comment by chrisbobbe** [comment]
Oh, interesting. From reading the [Apple doc I linked to](https://developer.apple.com/documentation/authenticationservices/asauthorizationappleidbutton), I wouldn't have guessed that Apple lets you create a custom button:

> Choose one of the built-in button styles and types, and change the corner radius of the button by setting the [cornerRadius](https://developer.apple.com/documentation/authenticationservices/asauthorizationappleidbutton/3153030-cornerradius) property, but don’t otherwise modify the style of the button.

It seems very Apple not to mention, on that crucial bit of documentation for what looks like The Way to do Apple auth, that some custom buttons are allowed. It's not quite as far as self-contradictory documentation, which we've seen with Apple before, since the `ASAuthorizationAppleIDButton` doc seems carefully worded to avoid categorically saying that this is the only approved button to use.

Without having gone through the entire implementation, I do think that the Expo doc seems reasonably accurate as far as it describes the UI constraints imposed by this particular Apple widget, which it uses. Apple's "Using the System-Provided Buttons" in the doc you linked to seems to correspond well with Expo's warnings and guidance.

I wonder if Expo ran into a lot of resistance from Apple on their submissions for review, as they were developing this (certainly reasonable, I'd say). That would explain Expo's laser-focus in their documentation on what Apple *doesn't* allow, as a good-willed effort to help developers avoid App store rejections. If this is the case, I might also expect that Expo's implementation is, in fact, optimized for accuracy — at least, accuracy in the correct use of this widget, and any other aspects of an Apple auth implementation that have tight published guidelines. It doesn't sound like it would need to be a very substantial wrapper, but I haven't looked very deeply.

Still, it would be most correct if Expo also mentioned that other UI choices are allowed (just that they introduce risk of rejection). I guess we could eventually use our own custom button instead of Expo's wrapper (probably starting from the custom button examples at the doc you linked to), and if we do, we'd probably still be able to rely on Expo for the non-graphical stuff (checking the status of credentials, actually signing in, etc.).

**Comment by chrisbobbe** [comment]
> I think I prefer the button styles that look most like the other buttons.

Maybe, the "WHITE_OUTLINE" in day mode and "BLACK" in night mode? 😛 The filenames in that image with the six examples shows you which is which.

**Comment by chrisbobbe** [comment]
> Automatic translation of the button's title into the language specified by the device

Looks like the Apple-provided button translates based on the device-level language choice. (Do we have a default language selection that's based on the device language? Might be worth an issue.)

**Comment by gnprice** [comment]
> > I think I prefer the button styles that look most like the other buttons.
> 
> Maybe, the "WHITE_OUTLINE" in day mode and "BLACK" in night mode?

Yep, those are the ones I had in mind :slightly_smiling_face: 


> (Do we have a default language selection that's based on the device language? Might be worth an issue.)

Yeah, this is #2612 .

**Comment by chrisbobbe** [comment]
(removed "blocked on other work" because Unimodules landed in #4016)
